### PR TITLE
Add intergration tests for tokenfatoty module

### PR DESCRIPTION
**Pull request by hieuvubk**

### CODE REVIEW

File: build/simd
Comment (mattverse) [inline build/simd:1]: Do we need to have this binary? This should probably be included in gitignore as well right?

File: dockernet/config.sh [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+VAL_MNEMONIC_4="tail forward era width glory magnet knock shiver cup broken turkey upgrade cigar story agent lake transfer misery sustain fragile parrot also air document"
+VAL_MNEMONIC_5="crime lumber parrot enforce chimney turtle wing iron scissors jealous indicate peace empty game host protect juice submit motor cause second picture nuclear area"
+VAL_MNEMONICS=(
+    "$VAL_MNEMONIC_1"
+    "$VAL_MNEMONIC_2"
+    "$VAL_MNEMONIC_3"
+    "$VAL_MNEMONIC_4"
+    "$VAL_MNEMONIC_5"
+)
+REV_MNEMONIC="tonight bonus finish chaos orchard plastic view nurse salad regret pause awake link bacon process core talent whale million hope luggage sauce card weasel"
```
Comment (mattverse) [inline dockernet/config.sh]: what is rev?

File: dockernet/pkill.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+pkill -f "docker-compose .*simd.* logs" | true
+pkill -f "/bin/bash.*create_logs.sh" | true
+pkill -f "tail .*.log" | true
```
Comment (mattverse) [inline dockernet/pkill.sh]: Do we need logs? / Why do we need logs?

File: dockernet/src/init_chain.sh [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+CHAIN_ID=$(GET_VAR_VALUE    ${CHAIN}_CHAIN_ID)
+BINARY=$(GET_VAR_VALUE      ${CHAIN}_BINARY)
+MAIN_CMD=$(GET_VAR_VALUE    ${CHAIN}_MAIN_CMD)
+DENOM=$(GET_VAR_VALUE       ${CHAIN}_DENOM)
+RPC_PORT=$(GET_VAR_VALUE    ${CHAIN}_RPC_PORT)
+NUM_NODES=$(GET_VAR_VALUE   ${CHAIN}_NUM_NODES)
+NODE_PREFIX=$(GET_VAR_VALUE ${CHAIN}_NODE_PREFIX)
+VAL_PREFIX=$(GET_VAR_VALUE  ${CHAIN}_VAL_PREFIX)
+# THe host zone can optionally specify additional the micro-denom granularity
```
Comment (mattverse) [inline dockernet/src/init_chain.sh]:
```suggestion
# The host zone can optionally specify additional the micro-denom granularity
```

**Review by mattverse — COMMENTED** [review]
Hello @hieuvubk, I have done a quick skim of this, I think it looks good, would it be possible to add README that contains 
- structure of the testings that you have added and what each file does
- List of properties that we ensure testing through this?

Thanks!

File: go.mod [hunk 4]
```diff
@@ -145,5 +145,6 @@ require (
 // use cosmos-compatible protobufs
 replace (
 	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
+	github.com/syndtr/goleveldb => github.com/syndtr/goleveldb v1.0.1-0.20200815110645-5c35d600f0ca // indirect
```
Comment (mattverse) [inline go.mod]: what's the reason we needed bump go level db version?

File: build/simd
Comment (hieuvubk) [reply build/simd:1]: Yeah u right its should be ignored

File: go.mod (hunk 4 above)
Comment (hieuvubk) [reply go.mod:148]: There was an error:
`
tokenfactory % bash dockernet/tests/run_all_tests.sh 
Error: rpc error: code = InvalidArgument desc = failed to load state at height 6; version does not exist (latest height: 6): invalid request
`
So we bump to goleveldb version osmosis is using.

Comment (hieuvubk) [reply go.mod]: There was an error when we try query: 

`tokenfactory % bash dockernet/tests/run_all_tests.sh 
Error: rpc error: code = InvalidArgument desc = failed to load state at height 6; version does not exist (latest height: 6): invalid request`

So we bump to goleveldb version osmosis is using.

File: dockernet/config.sh (hunk 1 above)
Comment (anhductn2001) [reply dockernet/config.sh]: It's  revenue account mnemonic. We don't need it anymore in our test, so I'll remove it.

Comment (anhductn2001) [reply dockernet/config.sh:59]: It's revenue account mnemonic. We don't need it anymore in our tests, so I'll remove it.

File: dockernet/pkill.sh (hunk 2 above)
Comment (anhductn2001) [reply dockernet/pkill.sh]: I think Yes. 
Commands in a pkill.sh file via bash to prevent the exit code of 1 from stopping the make in Linux.
We need logs to make debugging easier during testing.

File: dockernet/README.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
+`/dockernet` contains infrastructure that is used for testing and development of the Stride protocol. The scripts here support docker-image based testing, some of which are heavily inspired by those used by Osmosis and Quicksilver (although there have been large deviations from the original implementations since). The relevant licenses are included here.
```
Comment (mattverse) [inline dockernet/README.md]:
```suggestion
`/dockernet` contains infrastructure that is used for testing. The scripts here support docker-image based testing.
```

File: dockernet/README.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+## Testing
+
+* The purpose is to ensure that the main functions of the tokenfactory operate perfectly on any chain. So we set up a simulated environment on dockernet to test.
+
+* The main files that need attention:
+    * dockernet/build.sh: build both simd and node binaries for docker
+    * dockernet/config.sh: contains config variables
+    * dockernet/tests/integration_tests.sh: contains both test logic here
```
Comment (mattverse) [inline dockernet/README.md]:
```suggestion
    * dockernet/tests/integration_tests.sh: contains test logics here
```

File: dockernet/README.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
+`/dockernet` contains infrastructure that is used for testing and development of the Stride protocol. The scripts here support docker-image based testing, some of which are heavily inspired by those used by Osmosis and Quicksilver (although there have been large deviations from the original implementations since). The relevant licenses are included here.
+
+## Testing
+
+* The purpose is to ensure that the main functions of the tokenfactory operate perfectly on any chain. So we set up a simulated environment on dockernet to test.
+
+* The main files that need attention:
+    * dockernet/build.sh: build both simd and node binaries for docker
```
Comment (mattverse) [inline dockernet/README.md]:
```suggestion
    * dockernet/build.sh: build simd that will be used in docker
```

File: dockernet/config.sh [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+SIMD_MINT_EPOCH_DURATION="20s"
+UNBONDING_TIME="240s"
+MAX_DEPOSIT_PERIOD="30s"
+VOTING_PERIOD="30s"
+INITIAL_ANNUAL_PROVISIONS="10000000000000.000000000000000000"
+
+# LSM Params
+LSM_VALIDATOR_BOND_FACTOR="250"
+LSM_GLOBAL_LIQUID_STAKING_CAP="0.25"
+LSM_VALIDATOR_LIQUID_STAKING_CAP="0.50"
```
Comment (mattverse) [inline dockernet/config.sh]:
```suggestion
```

File: dockernet/config.sh [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+# Logs
+SIMD_LOGS=$LOGS/simd.log
+TX_LOGS=$DOCKERNET_HOME/logs/tx.log
+KEYS_LOGS=$DOCKERNET_HOME/logs/keys.log
+
+# COIN TYPES
+# Coin types can be found at https://github.com/satoshilabs/slips/blob/master/slip-0044.md
+COSMOS_COIN_TYPE=118
+ETH_COIN_TYPE=60
+TERRA_COIN_TYPE=330
```
Comment (mattverse) [inline dockernet/config.sh]:
```suggestion
```

File: dockernet/config.sh [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+# CHAIN PARAMS
+BLOCK_TIME='1s'
+SIMD_HOUR_EPOCH_DURATION="90s"
+SIMD_DAY_EPOCH_DURATION="140s"
+SIMD_EPOCH_EPOCH_DURATION="35s"
+SIMD_MINT_EPOCH_DURATION="20s"
+UNBONDING_TIME="240s"
+MAX_DEPOSIT_PERIOD="30s"
+VOTING_PERIOD="30s"
+INITIAL_ANNUAL_PROVISIONS="10000000000000.000000000000000000"
```
Comment (mattverse) [inline dockernet/config.sh]:
```suggestion
```

File: dockernet/config.sh [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+# CHAIN PARAMS
+BLOCK_TIME='1s'
+SIMD_HOUR_EPOCH_DURATION="90s"
+SIMD_DAY_EPOCH_DURATION="140s"
+SIMD_EPOCH_EPOCH_DURATION="35s"
+SIMD_MINT_EPOCH_DURATION="20s"
+UNBONDING_TIME="240s"
+MAX_DEPOSIT_PERIOD="30s"
+VOTING_PERIOD="30s"
```
Comment (mattverse) [inline dockernet/config.sh]: Can we set these as the block settings? (Seems that right now they are not being used)

File: dockernet/src/init_chain.sh [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+    $cmd add-genesis-account ${val_addr} ${VAL_TOKENS}${DENOM}
+
+    # Copy over the provider simd validator keys to the provider (in the event
+    # that we are testing ICS)
+    if [[ $CHAIN == "GAIA" && -d $DOCKERNET_HOME/state/${SIMD_NODE_PREFIX}${i} ]]; then
+        simd_config=$DOCKERNET_HOME/state/${SIMD_NODE_PREFIX}${i}/config
+        host_config=$DOCKERNET_HOME/state/${NODE_PREFIX}${i}/config
+        cp ${simd_config}/priv_validator_key.json ${host_config}/priv_validator_key.json
+        cp ${simd_config}/node_key.json ${host_config}/node_key.json
+    fi
```
Comment (mattverse) [inline dockernet/src/init_chain.sh]: do we need this part?

File: dockernet/src/init_chain.sh [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+            cp ${STATE}/${node_name}/config/gentx/*.json ${STATE}/${MAIN_NODE_NAME}/config/gentx/
+        fi
+
+        # and add each validator's keys to the first state directory
+        echo "$val_mnemonic" | $MAIN_CMD keys add $val_acct --recover --keyring-backend=test &> /dev/null
+    fi
+done
+
+# add a staker account for integration tests
+# the account should live on both stride and the host chain
```
Comment (mattverse) [inline dockernet/src/init_chain.sh]:
```suggestion
```

File: dockernet/src/init_chain.sh [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+    sed -i -E 's|localhost|0.0.0.0|g' $app_toml
+
+    sed -i -E "s|chain-id = \"\"|chain-id = \"${CHAIN_ID}\"|g" $client_toml
+    sed -i -E "s|keyring-backend = \"os\"|keyring-backend = \"test\"|g" $client_toml
+    sed -i -E "s|node = \".*\"|node = \"tcp://localhost:$RPC_PORT\"|g" $client_toml
+
+    sed -i -E "s|\"stake\"|\"${DENOM}\"|g" $genesis_json 
+    sed -i -E "s|\"aphoton\"|\"${DENOM}\"|g" $genesis_json # ethermint default
+
+    # add a validator account
```
Comment (mattverse) [inline dockernet/src/init_chain.sh:89]: asking to understand: Why is it that for every node, we're adding val account?

File: dockernet/tests/run_all_tests.sh [hunk 15]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+
+# run test files
+INTEGRATION_TEST_FILE=${SCRIPT_DIR}/integration_tests.sh
+
+bash $INTEGRATION_TEST_FILE
```
Comment (mattverse) [inline dockernet/tests/run_all_tests.sh:7]: We only have one test for now so should be fine to remove this right?

File: dockernet/pkill.sh [hunk 16]
```diff
@@ -0,0 +1,3 @@
+pkill -f "docker-compose .*simd.* logs" | true
+pkill -f "/bin/bash.*create_logs.sh" | true
```
Comment (mattverse) [inline dockernet/pkill.sh]: Can we change it so that we only get logs when we run with special make command and have a seperated command for running with and without make commands?

Comment (mattverse) [inline dockernet/pkill.sh]: Can we move this file to dockernet?

File: dockernet/src/init_chain.sh [hunk 17]
```diff
@@ -0,0 +1,178 @@
+#!/bin/bash
+
```
Comment (mattverse) [inline dockernet/src/init_chain.sh:2]: Can we put all the other sh files 
(build.sh, config.sh, start_network.sh) in src file for sanity?

**Review by mattverse — COMMENTED** [review]
Nice, left some last review before we merge. Once these are resolved I think this should be good to merge!

**Comment by mattverse** [comment]
@hieuvubk Can you also make the workflow run for 
- every PR 
- merge on main?

**Comment by anhductn2001** [comment]
> @hieuvubk Can you also make the workflow run for
> 
> * every PR
> * merge on main?

The workflow is running on every PR and push on Main branch, ser.

File: dockernet/src/init_chain.sh (hunk 17 above)
Comment (anhductn2001) [reply dockernet/src/init_chain.sh:2]: Ok @mattverse !

File: dockernet/pkill.sh (hunk 16 above)
Comment (anhductn2001) [reply dockernet/pkill.sh]: I don't really understand, is it already in dockernet?

File: dockernet/src/init_chain.sh (hunk 12 above)
Comment (anhductn2001) [reply dockernet/src/init_chain.sh]: We don't need this, I going to remove this part.

**Comment by anhductn2001** [comment]
Can you open Github Action access for this repo @mattverse ?

File: dockernet/config.sh (hunk 11 above)
Comment (anhductn2001) [reply dockernet/config.sh]: I removed it!

File: dockernet/tests/run_all_tests.sh (hunk 15 above)
Comment (anhductn2001) [reply dockernet/tests/run_all_tests.sh:7]: I think we should leave it that way. Although there is only one test now, bui it will be easy to develop if we want to add more tests later.

File: dockernet/pkill.sh (hunk 16 above)
Comment (anhductn2001) [reply dockernet/pkill.sh]: Yes, I 'll add new make command with debug flag when we want get logs.

**Comment by mattverse** [comment]
It's already set as "Allow all actions and reusable workflow" 🤔

**Comment by anhductn2001** [comment]
Hmm, Let's see this PR: https://github.com/notional-labs/tokenfactory/pull/1 workflow work well. I'm not sure why it's not running here.

File: dockernet/src/init_chain.sh (hunk 14 above)
Comment (hieuvubk) [reply dockernet/src/init_chain.sh:89]: Yeah cause each node we need a vaildator.

**Comment by mattverse** [comment]
@hieuvubk @anhductn2001 Let's not think about workflow for now, I think we can get this merged once we solve https://github.com/osmosis-labs/tokenfactory/pull/2#discussion_r1385935735 this comment!

**Comment by anhductn2001** [comment]
After looking again, I found that logging the state of the output node is important for dockernet, and we will not generate any other logs with create_log.sh so I will remove the kill process from it.

**Comment by mattverse** [comment]
@anhductn2001 What do you mean? Can you please elaborate more? We can still make it option to generate logs right? 👀

File: dockernet/src/start_chain.sh [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+$DOCKER_COMPOSE up -d $nodes_names
+
+$DOCKER_COMPOSE logs -f ${node_prefix}1 | sed -r -u "s/\x1B\[([0-9]{1,3}(;[0-9]{1,2})?)?[mGK]//g" > $log_file 2>&1 &
+
+printf "Waiting for Simd to start..."
+
+node_prefix=$(GET_VAR_VALUE SIMD_NODE_PREFIX)
+log_file=$DOCKERNET_HOME/logs/${node_prefix}.log
+
+( tail -f -n0 $log_file & ) | grep -q "finalizing commit of block"
```
Comment (anhductn2001) [inline dockernet/src/start_chain.sh:24]: Like here, I use log to check whether the node has committed the block or not @mattverse

File: dockernet/src/init_chain.sh [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+    node_name="${NODE_PREFIX}${i}"
+    # Moniker is of the form: SIMD_1
+    moniker=$(printf "${NODE_PREFIX}_${i}" | awk '{ print toupper($0) }')
+
+    # Create a state directory for the current node and initialize the chain
+    mkdir -p $STATE/$node_name
+    
+    # If the chains commands are run only from docker, grab the command from the config
+    # Otherwise, if they're run locally, append the home directory
+    cmd="$BINARY --home ${STATE}/$node_name"
```
Comment (anhductn2001) [inline dockernet/src/init_chain.sh:60]: State is used for init chain.

**Comment by anhductn2001** [comment]
No, we won't make options anymore

**Comment by mattverse** [comment]
Why not? 👀

**Comment by mattverse** [comment]
I do generally agree with the importance of logs, but questioning if we want logs being generated everytime we run test in our local machine. I was thinking of logs options like our current e2e tests, so that we can use logs if needed. Wdyt?
